controllers/metric_controller.go

```diff
@@ -208,49 +207,145 @@ func (r *MetricReconciler) collectMetrics(ctx context.Context, t *redskyv1beta1.
 
-func (r *MetricReconciler) target(ctx context.Context, namespace string, m *redskyv1beta1.Metric) (runtime.Object, error) {
-	switch m.Type {
-	case redskyv1beta1.MetricPods:
-		// Use the selector to get a list of pods
-		target := &corev1.PodList{}
-		sel, err := meta.MatchingSelector(m.Selector)
-		if err != nil {
-			return nil, err
-		}
+// collectionAttempt updates the status of the trial based on the outcome of an attempt to collect metric values.
+func (r *MetricReconciler) collectionAttempt(ctx context.Context, log logr.Logger, t *redskyv1beta1.Trial, v *redskyv1beta1.Value, probeTime *metav1.Time, err error) (*ctrl.Result, error) {
+	// Do not count retries against the remaining attempts
+	if merr, ok := err.(*metric.CaptureError); ok && merr.RetryAfter > 0 {
+		return &ctrl.Result{RequeueAfter: merr.RetryAfter}, nil
+	}
 
-		if err = r.List(ctx, target, client.InNamespace(namespace), sel); err != nil {
-			return nil, err
-		}
+	// Update the probe time and ensure that trial observed is still explicitly false (i.e. we have started observation but it is not complete)
+	trial.ApplyCondition(&t.Status, redskyv1beta1.TrialObserved, corev1.ConditionFalse, "", "", probeTime)
 
-		return target, nil
-	case redskyv1beta1.MetricPrometheus, redskyv1beta1.MetricJSONPath:
-		// Both Prometheus and JSONPath target a service
-		target := &corev1.ServiceList{}
+	// If there was only one attempt remaining, mark the trial as failed
+	if v.AttemptsRemaining == 1 {
+		trial.ApplyCondition(&t.Status, redskyv1beta1.TrialFailed, corev1.ConditionTrue, "MetricFailed", err.Error(), probeTime)
+		v.AttemptsRemaining = 0
 
-		if m.URL != "" {
-			return target, nil
+		// Metric errors contain additional information which should be logged for debugging
+		if merr, ok := err.(*metric.CaptureError); ok {
+			log.Error(merr, "Metric collection failed", "address", merr.Address, "query", merr.Query, "completionTime", merr.CompletionTime)
 		}
+	} else if v.AttemptsRemaining > 0 {
+		// Decrement the attempts remaining
+		v.AttemptsRemaining--
+	}
 
-		// Default label selector to `app: prometheus` if not specified
-		labelSelector := m.Selector
-		if labelSelector == nil && m.Type == redskyv1beta1.MetricPrometheus {
-			labelSelector = &metav1.LabelSelector{
-				MatchLabels: map[string]string{"app": "prometheus"},
-			}
-		}
+	// Record the update
+	return controller.RequeueConflict(r.Update(ctx, t))
+}
 
-		sel, err := meta.MatchingSelector(labelSelector)
-		if err != nil {
-			return nil, err
-		}
+// target looks up the Kubernetes object (if any) associated with a metric.
+func (r *MetricReconciler) target(ctx context.Context, t *redskyv1beta1.Trial, m *redskyv1beta1.Metric) (runtime.Object, error) {
+	if m.Type != redskyv1beta1.MetricKubernetes && m.Type != "" {
+		return nil, nil
+	}
+
+	// Get the target reference, default to the trial itself if there is no reference
+	targetRef := m.Target
+	if m.Target == nil || m.Target.Kind == "" {
+		return t, nil
+	}
```

Comment:
more of a nit, but do we want to do the assignment after the test or change the test to use the assignment ( targetRef ).

Reply:
We actually don't even need `targetRef` anymore.